receiver/saphanareceiver/client.go

```diff
@@ -0,0 +1,234 @@
+	return w.rows.Next()
+}
+
+func (w *standardResultWrapper) Scan(dest ...interface{}) error {
+	return w.rows.Scan(dest...)
+}
+
+func (w *standardResultWrapper) Close() error {
+	return w.rows.Close()
+}
+
+type standardDBWrapper struct {
+	db *sql.DB
+}
+
+func (w *standardDBWrapper) Close() error {
+	return w.db.Close()
+}
+
+func (w *standardDBWrapper) PingContext(ctx context.Context) error {
+	return w.db.PingContext(ctx)
+}
+
+func (w *standardDBWrapper) QueryContext(ctx context.Context, query string) (resultWrapper, error) {
+	rows, err := w.db.QueryContext(ctx, query)
+	if err != nil {
+		return nil, err
+	}
+	resultWrapper := standardResultWrapper{rows}
+	return &resultWrapper, nil
+}
+
+// Wraps the creation of a sqlDB so that it can be mocked in tests
+type sapHanaConnectionFactory interface {
+	getConnection(c driver.Connector) dbWrapper
+}
+
+type defaultConnectionFactory struct{}
+
+func (f *defaultConnectionFactory) getConnection(c driver.Connector) dbWrapper {
+	wrapper := standardDBWrapper{db: sql.OpenDB((c))}
+	return &wrapper
+}
+
+// Wraps a SAP HANA database connection, implements `client` interface.
+type sapHanaClient struct {
+	receiverConfig    *Config
+	connectionFactory sapHanaConnectionFactory
+	client            dbWrapper
+}
+
+var _ client = (*sapHanaClient)(nil)
+
+// Creates a SAP HANA database client
+func newSapHanaClient(cfg *Config, factory sapHanaConnectionFactory) client {
+	return &sapHanaClient{
+		receiverConfig:    cfg,
+		connectionFactory: factory,
+	}
+}
+
+func (c *sapHanaClient) Connect(ctx context.Context) error {
+	connector, err := sapdriver.NewDSNConnector(fmt.Sprintf("hdb://%s:%s@%s", c.receiverConfig.Username, c.receiverConfig.Password, c.receiverConfig.TCPAddr.Endpoint))
+	if err != nil {
+		return fmt.Errorf("error generating DSN for SAP HANA connection: %w", err)
+	}
+
+	tls, err := c.receiverConfig.TLSClientSetting.LoadTLSConfig()
+	if err != nil {
+		return fmt.Errorf("error generating TLS config for SAP HANA connection: %w", err)
+	}
+	connector.SetTLSConfig(tls)
+
+	connector.SetApplicationName("OpenTelemetry Collector")
+	client := c.connectionFactory.getConnection(connector)
+
+	err = client.PingContext(ctx)
+	if err == nil {
+		c.client = client
+	} else {
+		client.Close()
+	}
+
+	return err
+}
+
+func (c *sapHanaClient) Close() error {
+	if c.client != nil {
+		client := c.client
+		c.client = nil
+		return client.Close()
+	}
+	return nil
+}
+
+func (c *sapHanaClient) collectDataFromQuery(ctx context.Context, query *monitoringQuery) ([]map[string]string, error) {
+	rows, err := c.client.QueryContext(ctx, query.query)
+	if err != nil {
+		return nil, err
+	}
+	defer rows.Close()
+
+	errors := scrapererror.ScrapeErrors{}
+	data := []map[string]string{}
+
+ROW_ITERATOR:
+	for rows.Next() {
+		rowFields := make([]interface{}, 0)
+
+		// Build a list of addresses that rows.Scan will load column data into
+		for range query.orderedResourceLabels {
```

Comment:
Perhaps instead of iterating here, as well `orderedMetricLabels` and `orderedStats` below, rowFields could be created with single make? (since we know the sizes of each)

Reply:
rowFields needs to be a set of pointers to objects, so it needs more than just a singular make in my understanding, but I was able to update it to be a little bit cleaner.